hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/util/S3Utils.java

```diff
@@ -131,16 +136,69 @@ public static WebApplicationException wrapOS3Exception(OS3Exception ex) {
             .build());
   }
 
-  public static boolean hasSignedPayloadHeader(HttpHeaders headers) {
-    final String signingAlgorithm = headers.getHeaderString(X_AMZ_CONTENT_SHA256);
-    if (signingAlgorithm == null) {
-      return false;
+  public static boolean hasUnsignedPayload(@Nonnull String amzContentSha256Header) {
+    Objects.requireNonNull(amzContentSha256Header);
+    return amzContentSha256Header.equals(UNSIGNED_PAYLOAD) ||
+        amzContentSha256Header.equals(STREAMING_UNSIGNED_PAYLOAD_TRAILER);
+  }
+
+  public static boolean hasMultiChunksPayload(@Nonnull String amzContentSha256Header) {
+    Objects.requireNonNull(amzContentSha256Header);
+    // Multiple chunk uploads
+    // - https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html
+    // - https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming-trailers.html
+    // Possible values
+    // - STREAMING-UNSIGNED-PAYLOAD-TRAILER
+    // - STREAMING-AWS4-HMAC-SHA256-PAYLOAD
+    // - STREAMING-AWS4-HMAC-SHA256-PAYLOAD-TRAILER
+    // - STREAMING-AWS4-ECDSA-P256-SHA256-PAYLOAD
+    // - STREAMING-AWS4-ECDSA-P256-SHA256-PAYLOAD-TRAILER
+    // Currently since all the multi chunks values have x-amz-content-sha256 header value that starts
+    // with STREAMING, we can use this prefix to differentiates between multi chunks and single chunks upload.
+    // In the future if there are more multi chunks signature algorithms that has the same prefix,
+    // this function will be able to handle detect it.
+    return amzContentSha256Header.startsWith(MULTI_CHUNKS_UPLOAD_PREFIX);
+  }
+
+  public static void validateMultiChunksUpload(HttpHeaders headers, String amzDecodedContentLength,
+                                               String resource) throws OS3Exception {
+    final String contentEncoding = headers.getHeaderString(HttpHeaders.CONTENT_ENCODING);
+    // "Content-Encoding : aws-chunked" seems to only be sent for SDK V2, so ignore if there is no
+    // Content-Encoding header
+    if (contentEncoding == null) {
+      return;
```

Comment:
I'm wondering if using a plain return here is a good idea. It might allow some requests that should have a Content-Encoding header to pass validation.

Even though I assume the request will eventually fail when the InputStream is created, I think it would be better to catch this earlier by validating it here.

Reply:
Thanks for your review. Updated so that the following "x-amz-decoded-content-length" is still checked regardless whether "Content-Encoding" header is specified.